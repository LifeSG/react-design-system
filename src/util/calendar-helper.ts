import dayjs, { Dayjs } from "dayjs";

export namespace CalendarHelper {
    // =============================================================================
    // HELPER FUNCTIONS
    // =============================================================================
    export const generateDays = (calendarDate: Dayjs): Dayjs[][] => {
        const firstDayOfTheMonth = calendarDate.startOf("month");

        const firstDayOfFirstWeekOfMonth =
            dayjs(firstDayOfTheMonth).startOf("week");

        const firstDayOfEachWeek = generateFirstDayOfEachWeek(
            firstDayOfFirstWeekOfMonth
        );

        return firstDayOfEachWeek.map((date) => generateWeek(date));
    };

    export const generateMonths = (calendarDate: Dayjs): Dayjs[] => {
        const months: Dayjs[] = [];

        for (let i = 0; i < 12; i++) {
            const monthForSelectedDay = calendarDate.month(i);
            months.push(dayjs(monthForSelectedDay));
        }

        return months;
    };

    export const generateDecadeOfYears = (calendarDate: Dayjs): Dayjs[] => {
        const year = calendarDate.year();
        const decade = Math.floor(year / 10) * 10;

        const base = calendarDate.year(decade);
        const prev = base.subtract(1, "year");

        const years = [prev, base];

        for (let i = 1; i < 11; i++) {
            years.push(base.add(i, "year"));
        }

        return years;
    };

    export const getStartEndDecade = (calendarDate: Dayjs) => {
        const beginDecade = Math.floor(+calendarDate.format("YYYY") / 10) * 10;
        const endDecade = +dayjs(`${beginDecade + 9}-01-01`).format("YYYY");

        return {
            beginDecade,
            endDecade,
        };
    };
}

const generateFirstDayOfEachWeek = (day: Dayjs): Dayjs[] => {
    const dates: Dayjs[] = [day];

    for (let i = 1; i < 6; i++) {
        const date = day.add(i, "week");
        dates.push(date);
    }

    return dates;
};

const generateWeek = (day: Dayjs): Dayjs[] => {
    const dates: Dayjs[] = [];

    for (let i = 0; i < 7; i++) {
        const date = day.add(i, "day");
        dates.push(date);
    }

    return dates;
};
