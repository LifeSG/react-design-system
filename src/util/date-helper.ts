export namespace DateHelper {
    export const clampDay = (
        day: string,
        month: string,
        year: string
    ): string => {
        const dayNumber = parseInt(day);
        const monthNumber = parseInt(month);
        const yearNumber = parseInt(year);

        if (dayNumber == 0) {
            return "1";
        }

        switch (monthNumber) {
            case 1:
            case 3:
            case 5:
            case 7:
            case 8:
            case 10:
            case 12: {
                if (dayNumber > 31) {
                    return "31";
                }
                break;
            }
            case 4:
            case 6:
            case 9:
            case 11:
                if (dayNumber > 30) {
                    return "30";
                }
                break;
            case 2:
                if (isLeapYear(yearNumber)) {
                    if (dayNumber > 29) {
                        return "29";
                    }
                } else {
                    if (dayNumber > 28) {
                        return "28";
                    }
                }
                break;
            default:
                break;
        }

        return day;
    };
    export const clampMonth = (month: string): string => {
        const monthNumber = parseInt(month);

        if (monthNumber == 0) {
            return "1";
        }

        if (monthNumber > 12) {
            return "12";
        }
        return month;
    };
    export const isLeapYear = (year: number): boolean => {
        return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
    };
}
