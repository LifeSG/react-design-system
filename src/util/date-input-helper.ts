import dayjs from "dayjs";
import isBetween from "dayjs/plugin/isBetween";

dayjs.extend(isBetween);

export namespace DateInputHelper {
    const dateFormat = /^\d{4}-\d{2}-\d{2}$/; // YYYY-MM-DD

    export const isDateDisabled = (
        val: string | undefined,
        props: { disabledDates?: string[]; between?: [string, string] }
    ): boolean => {
        const { disabledDates, between } = props;
        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.includes(val)
        ) {
            return true;
        }

        if (between && between.length === 2) {
            const [min, max] = between;
            if (!min || !max) {
                console.warn('"between" prop is invalid');
                return true;
            }

            if (!dayjs(val).isBetween(min, max, "day", "[]")) {
                return true;
            }
        }

        return false;
    };

    export const validate = (
        startDate: string | undefined,
        endDate: string | undefined,
        disabledDates?: string[] | undefined,
        between?: string[] | undefined
    ): boolean => {
        if (!startDate || !endDate) {
            return false;
        }

        if (dayjs(startDate).isAfter(endDate)) {
            return false;
        }

        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.some((value) => [startDate, endDate].includes(value))
        ) {
            return false;
        }

        if (between && between.length) {
            if (
                ![startDate, endDate].every((selectedDate) =>
                    dayjs(selectedDate).isBetween(
                        between[0],
                        between[1],
                        "day",
                        "[]"
                    )
                )
            ) {
                return false;
            }

            if (!between.every((value) => dateFormat.test(value))) {
                return false;
            }
        }

        return true;
    };

    export const validateSingle = (
        value: string,
        disabledDates?: string[] | undefined,
        between?: string[] | undefined
    ) => {
        if (value.length === 0) {
            return false;
        }

        if (
            disabledDates &&
            disabledDates.length &&
            disabledDates.some((disabledDate) => value === disabledDate)
        ) {
            return false;
        }

        if (
            between &&
            between.length &&
            !dayjs(value).isBetween(between[0], between[1], "day", "[]")
        ) {
            if (!dayjs(value).isBetween(between[0], between[1], "day", "[]")) {
                return false;
            }

            if (!between.every((value) => dateFormat.test(value))) {
                return false;
            }
        }

        return true;
    };

    export const getFormattedRawValue = (
        startValue: string,
        endValue?: string
    ): {
        start: string[];
        end: string[] | undefined;
    } => {
        const delimiter = "-";
        const [year, month, day] = startValue.split(delimiter);

        const start = [day, month, year];

        let end = undefined;
        if (endValue) {
            const [endYear, endMonth, endDay] = endValue.split(delimiter);
            end = [endDay, endMonth, endYear];
        }

        return {
            start,
            end,
        };
    };

    export const sleep = (ms: number, controller: AbortController) =>
        new Promise((resolve) => {
            const timeoutId = setTimeout(resolve, ms);

            controller.signal.addEventListener("abort", () => {
                clearTimeout(timeoutId);
            });
        });
}
