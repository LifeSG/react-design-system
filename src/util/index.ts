export * from "./calendar-helper";
export * from "./date-helper";
export * from "./date-input-helper";
export * from "./simple-id-generator";
export * from "./string-helper";
export * from "./use-container-query";
export * from "./use-event-listener";
export * from "./use-next-input-state";
export * from "./use-state-ref";
