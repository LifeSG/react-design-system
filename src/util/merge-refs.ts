import { MutableRefObject, Ref } from "react";
import { isReact19 } from "./version";

export const mergeRefs = <T>(...refs: Ref<T>[]) => {
    return (value: T | null) => {
        const cleanups: (() => void)[] = [];

        for (const ref of refs) {
            if (typeof ref === "function") {
                const cleanup = ref(value);
                const isCleanup = typeof cleanup === "function";
                cleanups.push(isCleanup ? cleanup : () => ref(null));
            } else {
                (ref as MutableRefObject<T | null>).current = value;
                cleanups.push(() => {
                    (ref as MutableRefObject<T | null>).current = null;
                });
            }
        }

        if (isReact19()) {
            return () => {
                for (const cleanup of cleanups) cleanup();
            };
        }
    };
};
