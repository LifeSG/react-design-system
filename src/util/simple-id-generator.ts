export namespace SimpleIdGenerator {
    /**
     * Generates a simple generator for element ids
     */
    export const generate = (): string => {
        return Math.random().toString(36).substring(2, 9);
    };
}
