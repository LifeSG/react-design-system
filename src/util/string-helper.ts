import { MaskAttributeProps } from "../masked-input";

export namespace StringHelper {
    export const transformWithSpaces = (
        value: string | any,
        groupLength: number
    ) => {
        const regex = "(.{" + groupLength + "})";
        return value
            .replace(/[^\dA-Z]/gi, "")
            .replace(new RegExp(regex, "g"), "$1 ")
            .trim();
    };

    /**
     * Pads single digit values with a zero and trims leading zero for
     * values of length more than 2
     * @param value the string value
     * @param retainSingleZero setting true will allow single zero value
     */
    export const padValue = (
        value: string,
        retainSingleZero?: boolean
    ): string => {
        if (value === "0") {
            return retainSingleZero ? value : value.padStart(2, "0");
        }

        if (value === "") {
            return value;
        }

        const trimmedValue = parseInt(value);
        return trimmedValue.toString().padStart(2, "0");
    };

    /**
     * Checks if the text is required to be truncated and broken into two lines based
     * on the available width of container. **Note: This is relevant to font size 18px or 1.125rem only **
     * @param text input text
     * @param widthOfElement element width in px
     */
    export const shouldTruncateToTwoLines = (
        text: string,
        widthOfElement: number
    ): boolean => {
        // This is arbitrary and based on tests to derive that every 9px increment ~= 1 char length.
        // But this applies to font size 18px or 1.125rem only
        const estimatedCharOnLine = Math.floor(widthOfElement / 9);
        return (
            text.length >= estimatedCharOnLine * 2 ||
            (text.split(" ").length === 1 && text.length > estimatedCharOnLine)
        );
    };

    /**
     * Performs a truncation by adding ellipsis in the middle of the text.
     * @param text input text
     * @param widthOfElement the dynamic width of the container in px
     * @param minDivSize the minimum container size (using size on mobileS viewport as baseline)
     * @param baselineCharLength the baseline amount of characters to be displayed before truncation kicks. This
     * will be increased if there is more space available (derived from `widthOfElement`)
     * @param incrementFactor the size (in px).
     */
    export const truncateOneLine = (
        text: string,
        widthOfElement: number,
        minDivSize: number,
        baselineCharLength: number,
        incrementFactor = 8 // Arbitrary based on tests and applies to font size of 18px or 1.125rem
    ): string => {
        let additionalCharAllowed = 0;
        if (widthOfElement > minDivSize) {
            additionalCharAllowed = Math.floor(
                (widthOfElement - minDivSize) / incrementFactor
            );
        }
        const allowedCharLength = baselineCharLength + additionalCharAllowed;
        if (allowedCharLength < text.length) {
            const thresholdIndex = Math.floor(allowedCharLength / 2); // We will cut it halfway to derive x num of characters
            return (
                text.substring(0, thresholdIndex) + // Use the first x
                " ... " +
                text.substring(text.length - thresholdIndex, text.length) // Use the last x
            );
        }

        return text;
    };

    /**
     * Returns the width of the text in the specified font
     * https://stackoverflow.com/questions/118241/calculate-text-width-with-javascript/21015393#21015393
     */
    export const getTextWidth = (text: string, font: string): number => {
        if (typeof document === "undefined") {
            return 0;
        }

        // create a reusable canvas object for better performance
        const canvas =
            (getTextWidth as any).canvas ||
            ((getTextWidth as any).canvas = document.createElement("canvas"));
        const context = canvas.getContext("2d");
        context.font = font;

        const metrics = context.measureText(text);
        return metrics.width;
    };

    export const maskValue = (value: string, options?: MaskAttributeProps) => {
        if (!value) {
            return value;
        }

        const {
            maskRange,
            unmaskRange,
            maskChar = "â€¢",
            maskRegex,
            maskTransformer,
        } = options;

        if (maskTransformer) {
            return maskTransformer(value);
        } else if (maskRegex) {
            return value.replace(maskRegex, maskChar);
        } else if (maskRange) {
            const { startIndex, endIndex } = determineStartAndEndIndex(
                maskRange[0],
                maskRange[1]
            );
            return (
                value.substring(0, startIndex) +
                maskChar.repeat(
                    value.substring(startIndex, endIndex + 1).length
                ) +
                value.substring(endIndex + 1)
            );
        } else if (unmaskRange) {
            const { startIndex, endIndex } = determineStartAndEndIndex(
                unmaskRange[0],
                unmaskRange[1]
            );
            return (
                maskChar.repeat(value.substring(0, startIndex).length) +
                value.substring(startIndex, endIndex + 1) +
                maskChar.repeat(value.substring(endIndex + 1).length)
            );
        }

        return value;
    };

    const determineStartAndEndIndex = (index0: number, index1: number) => {
        return index0 < index1
            ? { startIndex: index0, endIndex: index1 }
            : { startIndex: index1, endIndex: index0 };
    };
}
