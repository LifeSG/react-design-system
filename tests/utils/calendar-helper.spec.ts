import dayjs, { Dayjs } from "dayjs";
import { CalendarHelper } from "../../src/util/calendar-helper";

describe("CalendarHelper", () => {
    describe("generateDays", () => {
        it("should generate 6 weeks from the specified month", () => {
            const date = dayjs("2023-01-10");

            const values = CalendarHelper.generateDays(date);

            expect(values).toHaveLength(6);
            expect(values).toMatchSnapshot();
        });

        it("should generate Feb 29 in a leap year", () => {
            const date = dayjs("2028-02-01");
            const leapDay = "2028-02-29";

            const values = CalendarHelper.generateDays(date);

            expect(containsDate(values.flat(), leapDay)).toBeTruthy();
        });

        it("should generate Feb 28 in a non leap year", () => {
            const date = dayjs("2027-02-01");
            const leapDay = "2028-02-29";

            const values = CalendarHelper.generateDays(date);

            expect(containsDate(values.flat(), leapDay)).toBeFalsy();
        });
    });

    describe("generateMonths", () => {
        it("should generate the date for each month in the specified year", () => {
            const date = dayjs("2023-02-01");
            const correctLength = 12;

            const months = CalendarHelper.generateMonths(date);

            expect(months).toHaveLength(correctLength);
            expect(months).toMatchSnapshot();
        });

        it("should generate Feb 29 in a leap year", () => {
            const date = dayjs("2024-03-31");
            const leapDay = "2024-02-29";

            const values = CalendarHelper.generateMonths(date);

            expect(containsDate(values, leapDay)).toBeTruthy();
        });

        it("should generate the last day of the month if it has less days than the specified month", () => {
            const date = dayjs("2023-03-31");
            const lastDayOfFeb = "2023-02-28";
            const lastDayOfApr = "2023-04-30";

            const values = CalendarHelper.generateMonths(date);

            expect(containsDate(values, lastDayOfFeb)).toBeTruthy();
            expect(containsDate(values, lastDayOfApr)).toBeTruthy();
        });
    });

    describe("generateDecadeOfYears", () => {
        it("should generate the date for the decade and 2 years before", () => {
            const date = dayjs("2023-04-01");
            const correctLength = 12;

            const values = CalendarHelper.generateDecadeOfYears(date);

            expect(values).toHaveLength(correctLength);
            expect(values[0].year()).toEqual(2019);
            expect(values[11].year()).toEqual(2030);
            expect(values).toMatchSnapshot();
        });

        it("should return Feb 28 in non leap years", () => {
            const date = dayjs("2024-02-29");
            const lastDayOfTheMonth = "2023-02-28";

            const values = CalendarHelper.generateDecadeOfYears(date);

            expect(containsDate(values, lastDayOfTheMonth)).toBeTruthy();
        });
    });

    describe("getStartEndDecade", () => {
        it("should return correct start year and end year in this decade", () => {
            const startYear = 2020;
            const endYear = 2029;
            const date = dayjs("2024-01-01");

            const { beginDecade, endDecade } =
                CalendarHelper.getStartEndDecade(date);

            expect(startYear).toEqual(beginDecade);
            expect(endYear).toEqual(endDecade);
        });
    });

    describe("isWithinRange", () => {
        it("should return true if the date is within range", () => {
            const date = dayjs("2023-01-04");
            const minDate = dayjs("2023-01-01");
            const maxDate = dayjs("2023-02-01");

            const isWithinRange = CalendarHelper.isWithinRange(
                date,
                minDate,
                maxDate
            );

            expect(isWithinRange).toBe(true);
        });

        it("should return true if the date is within range (inclusive check)", () => {
            const minDate = dayjs("2023-01-01");
            const maxDate = dayjs("2023-02-01");

            const sameAsMaxCheck = CalendarHelper.isWithinRange(
                maxDate,
                minDate,
                maxDate
            );
            const sameAsMinCheck = CalendarHelper.isWithinRange(
                minDate,
                minDate,
                maxDate
            );

            expect(sameAsMaxCheck).toBe(true);
            expect(sameAsMinCheck).toBe(true);
        });

        it("should return false if the date is out of range", () => {
            const date = dayjs("2023-03-04");
            const minDate = dayjs("2023-01-01");
            const maxDate = dayjs("2023-02-01");

            const isWithinRange = CalendarHelper.isWithinRange(
                date,
                minDate,
                maxDate
            );

            expect(isWithinRange).toBe(false);
        });

        it("should return true if date is same or after minDate (if only minDate provided)", () => {
            const date = dayjs("2023-01-02");
            const minDate = dayjs("2023-01-02");

            const isWithinRange = CalendarHelper.isWithinRange(date, minDate);
            const isMonthWithinRange = CalendarHelper.isWithinRange(
                date,
                minDate,
                undefined,
                "month"
            );

            expect(isWithinRange).toBe(true);
            expect(isMonthWithinRange).toBe(true);
        });

        it("should return true if date is same or before maxDate (if only maxDate provided)", () => {
            const date = dayjs("2023-01-02");
            const maxDate = dayjs("2023-01-02");

            const isWithinRange = CalendarHelper.isWithinRange(
                date,
                undefined,
                maxDate
            );

            const isMonthWithinRange = CalendarHelper.isWithinRange(
                date,
                undefined,
                maxDate,
                "month"
            );

            expect(isWithinRange).toBe(true);
            expect(isMonthWithinRange).toBe(true);
        });
    });

    describe("isPreviousMonthWithinRange", () => {
        it("should return true if the previous month of date is within minDate", () => {
            const prevMonth = dayjs("2023-04-15");
            const currentMonth = dayjs("2023-05-01");
            const nextMonth = dayjs("2023-06-15");
            const minDate = dayjs("2023-05-15");

            expect(
                CalendarHelper.isPreviousMonthWithinRange(prevMonth, minDate)
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousMonthWithinRange(currentMonth, minDate)
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousMonthWithinRange(nextMonth, minDate)
            ).toBe(true);
        });
    });

    describe("isPreviousYearWithinRange", () => {
        it("should return true if the previous year of date is within minDate", () => {
            const prevYear = dayjs("2022-12-31");
            const currentYear = dayjs("2023-01-01");
            const nextYear = dayjs("2024-01-01");
            const minDate = dayjs("2023-05-15");

            expect(
                CalendarHelper.isPreviousYearWithinRange(prevYear, minDate)
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousYearWithinRange(currentYear, minDate)
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousYearWithinRange(nextYear, minDate)
            ).toBe(true);
        });
    });

    describe("isPreviousDecadeWithinRange", () => {
        it("should return true if the previous decade of date is within minDate", () => {
            const prevDecade = dayjs("2018-12-31");
            const currentDecade = dayjs("2019-05-15");
            const nextDecade = dayjs("2031-01-01");
            const minDate = dayjs("2024-05-15");

            expect(
                CalendarHelper.isPreviousDecadeWithinRange(prevDecade, minDate)
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousDecadeWithinRange(
                    currentDecade,
                    minDate
                )
            ).toBe(false);
            expect(
                CalendarHelper.isPreviousDecadeWithinRange(nextDecade, minDate)
            ).toBe(true);
        });
    });

    describe("isNextMonthWithinRange", () => {
        it("should return true if the next month of date is within maxDate", () => {
            const prevMonth = dayjs("2023-04-15");
            const currentMonth = dayjs("2023-05-01");
            const nextMonth = dayjs("2023-06-15");
            const maxDate = dayjs("2023-05-15");

            expect(
                CalendarHelper.isNextMonthWithinRange(prevMonth, maxDate)
            ).toBe(true);
            expect(
                CalendarHelper.isNextMonthWithinRange(currentMonth, maxDate)
            ).toBe(false);
            expect(
                CalendarHelper.isNextMonthWithinRange(nextMonth, maxDate)
            ).toBe(false);
        });
    });

    describe("isNextYearWithinRange", () => {
        it("should return true if the next year of date is within maxDate", () => {
            const prevYear = dayjs("2022-12-31");
            const currentYear = dayjs("2023-12-31");
            const nextYear = dayjs("2024-01-01");
            const maxDate = dayjs("2023-05-15");

            expect(
                CalendarHelper.isNextYearWithinRange(prevYear, maxDate)
            ).toBe(true);
            expect(
                CalendarHelper.isNextYearWithinRange(currentYear, maxDate)
            ).toBe(false);
            expect(
                CalendarHelper.isNextYearWithinRange(nextYear, maxDate)
            ).toBe(false);
        });
    });

    describe("isNextDecadeWithinRange", () => {
        it("should return true if the next decade of date is within maxDate", () => {
            const prevDecade = dayjs("2018-12-31");
            const currentDecade = dayjs("2030-05-15");
            const nextDecade = dayjs("2031-01-01");
            const maxDate = dayjs("2024-05-15");

            expect(
                CalendarHelper.isNextDecadeWithinRange(prevDecade, maxDate)
            ).toBe(true);
            expect(
                CalendarHelper.isNextDecadeWithinRange(currentDecade, maxDate)
            ).toBe(false);
            expect(
                CalendarHelper.isNextDecadeWithinRange(nextDecade, maxDate)
            ).toBe(false);
        });
    });
});

// =============================================================================
// TEST HELPER
// =============================================================================
function containsDate(dates: Dayjs[], expectedDate: string) {
    const formattedDates = dates.map((day) => day.format("YYYY-MM-DD"));
    return formattedDates.includes(expectedDate);
}
