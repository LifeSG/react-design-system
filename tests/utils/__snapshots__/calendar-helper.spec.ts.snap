// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHelper generateDays should generate 6 weeks from the specified month 1`] = `
[
  [
    "2023-01-01T00:00:00.000Z",
    "2023-01-02T00:00:00.000Z",
    "2023-01-03T00:00:00.000Z",
    "2023-01-04T00:00:00.000Z",
    "2023-01-05T00:00:00.000Z",
    "2023-01-06T00:00:00.000Z",
    "2023-01-07T00:00:00.000Z",
  ],
  [
    "2023-01-08T00:00:00.000Z",
    "2023-01-09T00:00:00.000Z",
    "2023-01-10T00:00:00.000Z",
    "2023-01-11T00:00:00.000Z",
    "2023-01-12T00:00:00.000Z",
    "2023-01-13T00:00:00.000Z",
    "2023-01-14T00:00:00.000Z",
  ],
  [
    "2023-01-15T00:00:00.000Z",
    "2023-01-16T00:00:00.000Z",
    "2023-01-17T00:00:00.000Z",
    "2023-01-18T00:00:00.000Z",
    "2023-01-19T00:00:00.000Z",
    "2023-01-20T00:00:00.000Z",
    "2023-01-21T00:00:00.000Z",
  ],
  [
    "2023-01-22T00:00:00.000Z",
    "2023-01-23T00:00:00.000Z",
    "2023-01-24T00:00:00.000Z",
    "2023-01-25T00:00:00.000Z",
    "2023-01-26T00:00:00.000Z",
    "2023-01-27T00:00:00.000Z",
    "2023-01-28T00:00:00.000Z",
  ],
  [
    "2023-01-29T00:00:00.000Z",
    "2023-01-30T00:00:00.000Z",
    "2023-01-31T00:00:00.000Z",
    "2023-02-01T00:00:00.000Z",
    "2023-02-02T00:00:00.000Z",
    "2023-02-03T00:00:00.000Z",
    "2023-02-04T00:00:00.000Z",
  ],
  [
    "2023-02-05T00:00:00.000Z",
    "2023-02-06T00:00:00.000Z",
    "2023-02-07T00:00:00.000Z",
    "2023-02-08T00:00:00.000Z",
    "2023-02-09T00:00:00.000Z",
    "2023-02-10T00:00:00.000Z",
    "2023-02-11T00:00:00.000Z",
  ],
]
`;

exports[`CalendarHelper generateDecadeOfYears should generate the date for the decade and 2 years before 1`] = `
[
  "2019-04-01T00:00:00.000Z",
  "2020-04-01T00:00:00.000Z",
  "2021-04-01T00:00:00.000Z",
  "2022-04-01T00:00:00.000Z",
  "2023-04-01T00:00:00.000Z",
  "2024-04-01T00:00:00.000Z",
  "2025-04-01T00:00:00.000Z",
  "2026-04-01T00:00:00.000Z",
  "2027-04-01T00:00:00.000Z",
  "2028-04-01T00:00:00.000Z",
  "2029-04-01T00:00:00.000Z",
  "2030-04-01T00:00:00.000Z",
]
`;

exports[`CalendarHelper generateMonths should generate the date for each month in the specified year 1`] = `
[
  "2023-01-01T00:00:00.000Z",
  "2023-02-01T00:00:00.000Z",
  "2023-03-01T00:00:00.000Z",
  "2023-04-01T00:00:00.000Z",
  "2023-05-01T00:00:00.000Z",
  "2023-06-01T00:00:00.000Z",
  "2023-07-01T00:00:00.000Z",
  "2023-08-01T00:00:00.000Z",
  "2023-09-01T00:00:00.000Z",
  "2023-10-01T00:00:00.000Z",
  "2023-11-01T00:00:00.000Z",
  "2023-12-01T00:00:00.000Z",
]
`;
